tests/utils/icu_locale_utils_tests.cpp

```diff
@@ -0,0 +1,318 @@
+
+// -----------------------------------------------------------------------------
+// --SECTION--                                                        test suite
+// -----------------------------------------------------------------------------
+
+
+TEST(icu_locale_utils_test_suite, test_get_locale_from_vpack) {
+
+  {
+    std::string config = R"({"language" : "de", "country" : "DE"})";
+    auto vpack = VPackParser::fromJson(config.c_str(), config.size());
+    icu::Locale actual_locale;
+    ASSERT_TRUE(get_locale_from_vpack(vpack->slice(), actual_locale));
+
+    icu::Locale expected_locale("de", "DE");
+    ASSERT_EQ(actual_locale, expected_locale);
+  }
+
+  {
+    std::string config = R"({"language" : "EN", "country" : "US", "variant" : "_phonebook"})";
+    auto vpack = VPackParser::fromJson(config.c_str(), config.size());
+    icu::Locale actual_locale;
+    ASSERT_TRUE(get_locale_from_vpack(vpack->slice(), actual_locale));
+
+    icu::Locale expected_locale("EN", "US", "phonebook");
+    ASSERT_EQ(actual_locale, expected_locale);
+  }
+
+  {
+    std::string config = R"({"language" : "EN", "country" : "EN", "variant" : "_pinyan"})";
+    auto vpack = VPackParser::fromJson(config.c_str(), config.size());
+    icu::Locale actual_locale;
+    ASSERT_TRUE(get_locale_from_vpack(vpack->slice(), actual_locale));
+
+    icu::Locale expected_locale("de", "DE", "phonebook");
+    ASSERT_NE(actual_locale, expected_locale);
+  }
+
+  {
+    std::string config = R"({"language" : "EN", "country" : "EN", "variant" : "_pinyan", "encoding" : "utf-16"})";
+    auto vpack = VPackParser::fromJson(config.c_str(), config.size());
+    icu::Locale actual_locale;
+    ASSERT_FALSE(get_locale_from_vpack(vpack->slice(), actual_locale));
+  }
+
+  {
+    std::string config = R"({"language" : "EN", "country" : "US", "variant" : "_pinyan", "encoding" : "utf-8"})";
+    auto vpack = VPackParser::fromJson(config.c_str(), config.size());
+    icu::Locale actual_locale;
+    ASSERT_TRUE(get_locale_from_vpack(vpack->slice(), actual_locale));
+
+    icu::Locale expected_locale("en", "US", "pinyan");
+    ASSERT_EQ(actual_locale, expected_locale);
+  }
+
+  {
+    std::string config = R"({"language" : "EN", "variant" : "_pinyan"})";
+    auto vpack = VPackParser::fromJson(config.c_str(), config.size());
+    icu::Locale actual_locale;
+    ASSERT_TRUE(get_locale_from_vpack(vpack->slice(), actual_locale));
+
+    icu::Locale expected_locale("en", NULL, "pinyan");
+    ASSERT_EQ(actual_locale, expected_locale);
+  }
+
+  {
+    std::string config = R"({"language" : "DE@collation=phonebook"})";
+    auto vpack = VPackParser::fromJson(config.c_str(), config.size());
+    icu::Locale actual_locale;
+    ASSERT_TRUE(get_locale_from_vpack(vpack->slice(), actual_locale));
+
+    icu::Locale expected_locale("DE", NULL, NULL, "collation=phonebook");
+    ASSERT_EQ(actual_locale, expected_locale);
+  }
+}
+
+TEST(icu_locale_utils_test_suite, get_locale_from_str) {
```

Comment:
Please add some tests  with totally malformed string to be sure we could safely handle such input (e.g. do not crash).
Like "@ru_De.utf-8" or ".@".